### DISCUSSION TITLE
How to use @noble/hashes in a contract?

### DISCUSSION THREAD

**Discussion by tgrecojs** [Agoric/agoric-sdk#9958] [Q&A] [answered]
I've been using the `@noble/hashes` library within my project for the purpose of working with merkle trees.

To get a little more specific about my usage of the library, I use it to:

1. construct a merkle tree containing a fixed set of public keys. Each public key corresponds to agoric wallet.
2. generate validity proofs for a particular public key that exists within the tree upon
3. reconstruct the root hash of the merkle tree using the validity proof. **This is the only operation that takes place on-chain**. The validity proof is supplied to the contract as a part of `offerArgs`.

I've been able to test the contract code and verify that it works in a controlled, simulated environment, however, I run into issues when attempting to deploy the contract for testing on the agoric blockchain.

The block below contains the errors that arise (seemingly when the contract instance is attempting to start up).

```

  SyntaxError {
    message: 'Possible import expression rejected at <unknown>:7. (SES_IMPORT_REJECTED)',
  }

  › Object.execute (.../compartment-mapper/src/parse-pre-cjs.js:30:27)
  › require (.../compartment-mapper/src/parse-cjs-shared-export-wrapper.js:135:29)
  › eval (eval at <anonymous> (eval at makeEvaluate (file:///Users/tgreco/dapp-ertp-airdrop/node_modules/ses/src/make-evaluate.js:92:27)), <anonymous>:5:20)
  › Object.execute (.../compartment-mapper/src/parse-pre-cjs.js:40:1)
  › require (.../compartment-mapper/src/parse-cjs-shared-export-wrapper.js:135:29)
  › eval (eval at <anonymous> (eval at makeEvaluate (file:///Users/tgreco/dapp-ertp-airdrop/node_modules/ses/src/make-evaluate.js:92:27)), <anonymous>:4:18)
  › Object.execute (.../compartment-mapper/src/parse-pre-cjs.js:40:1)
  › eval (.../agoric-basics-contract/src/merkle-tree/index.js:1:165)
  › eval (.../agoric-basics-contract/src/airdrop/airdropKitCreator.js:1:387)
  › async importBundle (.../import-bundle/src/index.js:57:18)
```

Having said all of this, i am opening up this discussion in an attempt to figure out the best way to figure out the best way forward for myself, as well as future applications which choose to ship atop agoric. While I've managed to work through these issues, I keep finding myself in situations where an issue I thought I had put to rest comes back alive and as one can expect, it has burdened the development process quite a bit. 

The noble-hashes library is rare in its lack of dependence on any external modules. I mention this as I would expect it be much less problematic than if I was attempting to use a library that did rely on other libraries (which most likely rely on other libraries...which most likely rely on other libraries...). You can view it for yourself [here](https://github.com/paulmillr/noble-hashes/blob/main/package.json).

The source code for this contract can be found [here](https://github.com/hindley-milner-systems/dapp-ertp-airdrop/blob/dev/contract/src/airdrop/airdropKitCreator.js). Changes are expected in this file, but the way in which this `getMerkeRootFromMerkleProof` function is used will not change.

Thanks! cc @dckc @Jovonni

**Comment by dckc** [comment]
What are the steps leading up  to the `SyntaxError`? How exactly would I reproduce it?

**Reply by tgrecojs** [reply]
below are the steps needed to reproduce the error i am mentioning.

### 1. set up a new project and install `@noble/hashes` within it
```
yarn create @agoric/dapp noble-hashes-app
cd noble-hashes-dapp/contract
# run test to confirm that the contract executes as expected
yarn run test
# add @noble/hashes to project
yarn add @noble/hashes
```
### 2. Add arbitrary code that makes use of some code from the `noble/hashes` library.
I've updated the `offer-up.contract.js` file with an import statement as well as some code that utilizes the imported function `sha256`. Editing `offer-up.contract.js` as below works fine.

```diff
$ git diff src
diff --git a/contract/src/offer-up.contract.js b/contract/src/offer-up.contract.js
index e24f7d7..47a5f29 100644
--- a/contract/src/offer-up.contract.js
+++ b/contract/src/offer-up.contract.js
@@ -24,6 +24,8 @@ import { M, getCopyBagEntries } from '@endo/patterns';
 import { AssetKind } from '@agoric/ertp/src/amountMath.js';
 import { AmountShape } from '@agoric/ertp/src/typeGuards.js';
 import { atomicRearrange } from '@agoric/zoe/src/contractSupport/atomicTransfer.js';
+import { sha256 } from '@noble/hashes/sha256';
+
 import '@agoric/zoe/exported.js';
 
 const { Fail, quote: q } = assert;
@@ -95,6 +97,8 @@ export const start = async zcf => {
     exit: M.any(),
   });
 
+  const createHash = input => sha256(input);
+
   /** a seat for allocating proceeds of sales */
   const proceeds = zcf.makeEmptySeatKit().zcfSeat;
```


### 3. re-run the test code and watch the error appear.
after saving the newly updated contract, re-run the test code to finish reproducing this error.
```
yarn run test
```

**Reply by dckc** [reply]
I've taken the liberty to promote `@noble/hashes` to the title of this discussion (and rephrased it as a question).

If there's a more general issue here, we can widen the title again.

**Reply by dckc** [reply]
I found it useful to look at the contract changes as a diff:

```diff
$ git diff src
diff --git a/contract/src/offer-up.contract.js b/contract/src/offer-up.contract.js
index e24f7d7..47a5f29 100644
--- a/contract/src/offer-up.contract.js
+++ b/contract/src/offer-up.contract.js
@@ -24,6 +24,8 @@ import { M, getCopyBagEntries } from '@endo/patterns';
 import { AssetKind } from '@agoric/ertp/src/amountMath.js';
 import { AmountShape } from '@agoric/ertp/src/typeGuards.js';
 import { atomicRearrange } from '@agoric/zoe/src/contractSupport/atomicTransfer.js';
+import { sha256 } from '@noble/hashes/sha256';
+
 import '@agoric/zoe/exported.js';
 
 const { Fail, quote: q } = assert;
@@ -95,6 +97,8 @@ export const start = async zcf => {
     exit: M.any(),
   });
 
+  const createHash = input => sha256(input);
+
   /** a seat for allocating proceeds of sales */
   const proceeds = zcf.makeEmptySeatKit().zcfSeat;
```

**Reply by dckc** [reply]
> I've taken the liberty to promote `@noble/hashes` to the title of this discussion (and rephrased it as a question).
> 
> If there's a more general issue here, we can widen the title again.

speaking of the more general issue:

 - https://github.com/Agoric/agoric-sdk/discussions/9986

**Reply by dckc** [reply, marked as answer]
I just repeated the experiment. It works now that I get 1.5.0 from `yarn add @noble/hashes`.

**Comment by erights** [comment]
See also https://github.com/endojs/endo/blob/master/packages/ses/error-codes/SES_IMPORT_REJECTED.md

**Reply by dckc** [reply]
@erights is it clear to you where the import syntax occurs?

When I hit `SES_IMPORT_REJECTED`, I expect to see the file and line number of the syntax error, but all I get is a stack trace of endo code that did the rejecting. It's like a compiler crash instead of a compiler reporting an error in the compiled code:

```
REJECTED from ava test("Start the contract"): (SyntaxError#2)
SyntaxError#2: Possible import expression rejected at <unknown>:7. (SES_IMPORT_REJECTED)
  at Object.execute (.../compartment-mapper/src/parse-pre-cjs.js:30:27)
  at require (.../compartment-mapper/src/parse-cjs-shared-export-wrapper.js:135:29)
  at eval (eval at <anonymous> (eval at makeEvaluate (file:///tmp/noble-hashes-app/node_modules/ses/src/make-evaluate.js:92:27)), <anonymous>:5:20)
  at Object.execute (.../compartment-mapper/src/parse-pre-cjs.js:40:1)
  at require (.../compartment-mapper/src/parse-cjs-shared-export-wrapper.js:135:29)
  at eval (eval at <anonymous> (eval at makeEvaluate (file:///tmp/noble-hashes-app/node_modules/ses/src/make-evaluate.js:92:27)), <anonymous>:4:18)
  at Object.execute (.../compartment-mapper/src/parse-pre-cjs.js:40:1)
  at eval (.../noble-hashes-app-contract/src/offer-up.contract.js:1:182)
  at async importBundle (.../import-bundle/src/index.js:57:18)
```

p.s. I opened an endo issue about this:

 - https://github.com/endojs/endo/issues/2430

**Comment by turadg** [comment]
@tgrecojs I think what you're encountering has been fixed upstream,

[Adjust typescript. Bump jsbt. Emit separate ts declarations for esm](https://github.com/paulmillr/noble-hashes/commit/1fddf3dd54b02190bf4c5d7629b2dfac88c75107)

It hasn't been released yet,
https://github.com/paulmillr/noble-hashes/compare/1.4.0...main

But I hear from the maintainer that a new release is in security review and may be cut by end of the month.

**Reply by tgrecojs** [reply]
thanks @turadg! 

I will check out the links i've added and let you know how i fare. i sincerely appreciate the input 👍 ❤️

**Reply by dckc** [reply]
I'm not able to confirm, @turadg . I used `yarn add https://github.com/paulmillr/noble-hashes#1fddf3dd54b02190bf4c5d7629b2dfac88c75107` but now I get:

```
REJECTED from ava test("Start the contract"): (Error#1)
Error#1: Cannot find file for internal module "./esm/sha256.js" (with candidates "./esm/sha256.js", "./esm/sha256.js.js", "./esm/sha256.js.json", "./esm/sha256.js.node", "./esm/sha256.js/index.js", "./esm/sha256.js/index.json", "./esm/sha256.js/index.node") in package file:///tmp/noble-hashes-app/contract/node_modules/@noble/hashes/
  at Object.execute (.../compartment-mapper/src/import-archive.js:61:7)
  at eval (.../noble-hashes-app-contract/src/offer-up.contract.js:1:182)
  at async importBundle (.../import-bundle/src/index.js:57:18)
```

@kriskowal any ideas?

p.s. To reproduce the original issue, I also did a `yarn install` before `yarn run test`. And I resolved the `-app` vs `-dapp` typo between the `yarn create` and `cd` steps.

**Reply by turadg** [reply]
Sourcing the repo won't work because it doesn't have the built assets. In agoric-sdk I resorted to building them locally and then patching: https://github.com/Agoric/agoric-sdk/blob/master/patches/%40noble%2Bhashes%2B1.4.0.patch

You could also make a fork, run the build and check those in.

**Reply by tgrecojs** [reply]
I managed to get things working but my pathway there was an odd one that consisted of my compiling the project, and then copying those files into my own repository's top-level `node_modules` folder.

I'm also seeing that the process of using the library in my project isn't as easy as cloning my repo and running yarn install, so I'm thinking that it's in my best interest to add the compiled version of `noble/hashes` to my repository... 🤔

**Reply by dckc** [reply]
> I'm thinking that it's in my best interest to add the compiled version of noble/hashes to my repository...

yes, that seems like a reasonable approach here.

I'd put it in something like `src/vendor/@noble/hashes`.
